### PR TITLE
feat(cudf): Add startswith expression support

### PR DESCRIPTION
**Pull request by firestarman**
## Summary
- register `startswith` in the cuDF function evaluator for Spark string expressions
- support `startswith(column, constant)`, `startswith(column, column)`, and `startswith(constant, column)` on the GPU path
- preserve Spark null semantics for column-pattern evaluation and keep `startswith(constant, constant)` off the GPU path until row-count support exists

## Test plan
- [x] `velox_cudf_expression_selection_test --gtest_filter='*Startswith*'`
- [x] `velox_cudf_spark_filter_project_test --gtest_filter='*startswith*'`
- [x] `velox_cudf_hash_join_test --gtest_filter='*innerJoinWithMixedFilterPrecomputation*:*leftJoinWithStringFunctionFilter*'`
- [x] `velox_cudf_table_scan_test --gtest_filter='TableScanTest.filterPushdown:TableScanTest.remainingFilterExtraction'`

### CODE REVIEW

**Review by bdice — APPROVED** [review]
Nice work! All seems fine to me.

File: velox/experimental/cudf/expression/ExpressionEvaluator.cpp [hunk 1]
```diff
@@ -1095,6 +1159,15 @@ bool FunctionExpression::canEvaluate(std::shared_ptr<velox::exec::Expr> expr) {
...
   }
 
+  if (
+      opName == "startswith" && expr->inputs().size() == 2 &&
+      expr->inputs()[0]->name() == "literal" &&
+      expr->inputs()[1]->name() == "literal") {
+    // The function path only sees subexpression columns, so a fully constant
+    // startswith has no input column to infer the output row count from.
+    return false;
+  }
```
Comment (devavret) [inline velox/experimental/cudf/expression/ExpressionEvaluator.cpp]: We don't want function specific logic to bleed out into canEvaluate checker. cast is a special case and treated differently in velox. Realistically, we'll never receive only constants because expression optimization will fold this constant before we reach here

**Review by devavret — CHANGES_REQUESTED** [review]
impl LGTM

File: velox/experimental/cudf/expression/ExpressionEvaluator.cpp (hunk 1 above)
Comment (firestarman) [reply velox/experimental/cudf/expression/ExpressionEvaluator.cpp]: Thx for review.
Good suggestion, removed this specific code block.

**Review by devavret — APPROVED** [review]

File: velox/experimental/cudf/expression/ExpressionEvaluator.cpp [hunk 2]
```diff
@@ -593,22 +594,74 @@ class StartswithFunction : public CudfFunction {
...
+    auto rowCount =
+        inputColumns.empty() ? vector_size_t{1} : asView(inputColumns[0]).size();
+
+    std::unique_ptr<cudf::column> inputColumnHolder;
+    cudf::column_view inputCol;
+    if (inputIsConstant_) {
+      cudf::string_scalar inputScalar(input_, !inputIsNull_, stream, mr);
+      inputColumnHolder =
+          cudf::make_column_from_scalar(inputScalar, rowCount, stream, mr);
+      inputCol = inputColumnHolder->view();
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/ExpressionEvaluator.cpp:1203]: The `inputIsNull_ == true` path is untested — no test calls `startswith(cast(null as varchar), c0)`. Would adding a test similar to `startswithConstantInput` but with a null constant input make sense? On GPU, silent null-handling bugs are expensive to debug later.

File: velox/experimental/cudf/expression/ExpressionEvaluator.cpp [hunk 3]
```diff
@@ -593,22 +594,74 @@ class StartswithFunction : public CudfFunction {
...
+      }
+      cudf::string_scalar patternScalar(pattern_, true, stream, mr);
+      return cudf::strings::starts_with(inputCol, patternScalar, stream, mr);
+    }
+
+    auto patternCol = asView(inputColumns[nextInput]);
+    auto result = cudf::strings::starts_with(inputCol, patternCol, stream, mr);
+    auto [nullMask, nullCount] =
+        cudf::bitmask_and(cudf::table_view({inputCol, patternCol}), stream, mr);
+    result->set_null_mask(std::move(nullMask), nullCount);
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/ExpressionEvaluator.cpp:1222]: The `bitmask_and` + `set_null_mask` after `starts_with(col, col)` — is this correcting a specific cuDF behavior, or a safety net? A one-line comment would help future readers decide whether this can be removed if cuDF changes its null semantics.

File: velox/experimental/cudf/expression/ExpressionEvaluator.cpp [hunk 4]
```diff
@@ -593,22 +594,74 @@ class StartswithFunction : public CudfFunction {
...
   ColumnOrView eval(
       std::vector<ColumnOrView>& inputColumns,
       rmm::cuda_stream_view stream,
       rmm::device_async_resource_ref mr) const override {
-    auto inputCol = asView(inputColumns[0]);
-    cudf::string_scalar patternScalar(pattern_, true, stream, mr);
-    return cudf::strings::starts_with(inputCol, patternScalar, stream, mr);
+    size_t nextInput = 0;
+    auto rowCount =
+        inputColumns.empty() ? vector_size_t{1} : asView(inputColumns[0]).size();
```
Comment (karthikeyann) [inline velox/experimental/cudf/expression/ExpressionEvaluator.cpp]: `inputColumns.empty() ? vector_size_t{1} : ...` is unreachable since the `VELOX_CHECK` on line 617 guarantees at least one non-constant arg. Could simplify to `asView(inputColumns[0]).size()`, or add a comment that this is defensive.

File: velox/experimental/cudf/tests/sparksql/FilterProjectTest.cpp [hunk 5]
```diff
@@ -118,6 +119,134 @@ TEST_F(CudfFilterProjectTest, dateAdd) {
...
+                  .setParseOptions(options_)
+                  .values({data})
+                  .project({"startswith(c0, c1) AS c2"})
+                  .planNode();
+
+  auto expected = makeRowVector({
+      makeNullableFlatVector<bool>(
+          {true, false, std::nullopt, std::nullopt, true, false, false}),
+  });
+  auto result = AssertQueryBuilder(plan).copyResults(pool());
```
Comment (karthikeyann) [inline velox/experimental/cudf/tests/sparksql/FilterProjectTest.cpp:200]: nitpick: uses `copyResults` + `assertEqualVectors` where `assertResults(expected)` would match the other six startswith tests.

**Review by karthikeyann — APPROVED** [review]
Thank you for the contribution. 
Test suite is thorough. Three minor suggestions  below.

File: velox/experimental/cudf/expression/ExpressionEvaluator.cpp (hunk 2 above)
Comment (firestarman) [reply velox/experimental/cudf/expression/ExpressionEvaluator.cpp:1203]: nice catch. Will add one in later PRs

File: velox/experimental/cudf/tests/sparksql/FilterProjectTest.cpp (hunk 5 above)
Comment (firestarman) [reply velox/experimental/cudf/tests/sparksql/FilterProjectTest.cpp:200]: Agreed.
`assertResults(expected)` is cleaner here, and I'll switch it for consistency in later PRs.

File: velox/experimental/cudf/expression/ExpressionEvaluator.cpp (hunk 3 above)
Comment (firestarman) [reply velox/experimental/cudf/expression/ExpressionEvaluator.cpp:1222]: Good point.

I found the cuDF the column/column `starts_with` path does not fully preserve the target column's null semantics. In particular, a null target can come back as a valid `false`. 

I choose to follow the Spark semantics here, where `startswith(c0, c1)` should be null
if either input is null, so the extra `bitmask_and` / `set_null_mask` is intentional.

Will add the comment in later PRs.
